quil-rs/src/stats/mod.rs

```diff
@@ -0,0 +1,270 @@
+
+use std::collections::HashSet;
+use std::ops::Neg;
+
+use crate::{
+    instruction::{Instruction, Qubit},
+    Program,
+};
+use execution_graph::{Error as ExecutionGraphError, ExecutionGraph};
+
+pub trait InstructionsSource {
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_;
+    fn num_instructions(&self) -> usize;
+    fn get_used_qubits(&self) -> impl IntoIterator<Item = &Qubit>;
+}
+
+impl InstructionsSource for Program {
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_ {
+        Program::body_instructions(self)
+    }
+
+    fn num_instructions(&self) -> usize {
+        Program::num_instructions(self)
+    }
+
+    // There's no generic trait for "set-like" types, so we return a concrete
+    // type here (as an API convenience) but do not require this type in the
+    // trait definition.
+    #[allow(refining_impl_trait)]
+    fn get_used_qubits(&self) -> &HashSet<Qubit> {
+        Program::get_used_qubits(self)
+    }
+}
+
+impl InstructionsSource for &Program {
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_ {
+        (*self).body_instructions()
+    }
+
+    fn num_instructions(&self) -> usize {
+        (*self).num_instructions()
+    }
+
+    // See note on `Program` impl above.
+    #[allow(refining_impl_trait)]
+    fn get_used_qubits(&self) -> &HashSet<Qubit> {
+        (*self).get_used_qubits()
+    }
+}
+
+pub struct ProgramStats<S: InstructionsSource> {
+    source: S,
+    execution_graph: ExecutionGraph,
+}
+
+fn make_execution_graph<S: InstructionsSource>(
+    source: S,
+) -> Result<ExecutionGraph, ExecutionGraphError> {
+    ExecutionGraph::new(source.body_instructions().cloned())
+}
+
+impl<'a> ProgramStats<&'a Program> {
+    pub fn from_program(program: &'a Program) -> Result<Self, ExecutionGraphError> {
+        let execution_graph = make_execution_graph(program)?;
+
+        Ok(Self {
+            source: program,
+            execution_graph,
+        })
+    }
+
+    // Provided because Program specifically returns a hash-set, but in general
+    // `InstructionsSource` may not provide that feature.
+    pub fn set_of_qubits_used(&self) -> &HashSet<Qubit> {
+        self.source.get_used_qubits()
+    }
+}
+
+impl<S: InstructionsSource> ProgramStats<S> {
+    fn execution_graph(&self) -> &ExecutionGraph {
+        &self.execution_graph
+    }
+
+    /// The total number of instructions in the program *body*.
+    ///
+    /// This does not include:
+    /// - memory region definitions
+    /// - frame definitions
+    /// - waveform definitions
+    /// - gate definitions
+    pub fn body_instruction_count(&self) -> usize {
+        self.source.body_instructions().count()
+    }
+
+    /// The total number of instructions in the program.
+    ///
+    /// This includes all definitions excluded by [`Program::instruction_count`].
+    pub fn instruction_count(&self) -> usize {
+        self.source.num_instructions()
```

Comment:
Ignoring the comments about feelings about convention, we should unify the names, IMO.